src/cavegen.cpp

```diff
@@ -23,46 +23,174 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "mapgen_v5.h"
 #include "mapgen_v6.h"
 #include "mapgen_v7.h"
+#include "mg_biome.h"
 #include "cavegen.h"
 
 NoiseParams nparams_caveliquids(0, 1, v3f(150.0, 150.0, 150.0), 776, 3, 0.6, 2.0);
 
 
-///////////////////////////////////////// Caves V5
+////
+//// CavesNoiseIntersection
+////
 
+CavesNoiseIntersection::CavesNoiseIntersection(
+	INodeDefManager *nodedef, BiomeManager *biomemgr, v3s16 chunksize,
+	NoiseParams *np_cave1, NoiseParams *np_cave2, int seed, float cave_width)
+{
+	assert(nodedef);
+	assert(biomemgr);
+
+	m_ndef = nodedef;
+	m_bmgr = biomemgr;
+
+	m_csize = chunksize;
+	m_cave_width = cave_width;
 
-CaveV5::CaveV5(Mapgen *mg, PseudoRandom *ps)
+	m_ystride    = m_csize.X;
+	m_zstride_1d = m_csize.X * (m_csize.Y + 1);
+
+	// Noises are created using 1-down overgeneration
+	// A Nx-by-1-by-Nz-sized plane is at the bottom of the desired for
+	// re-carving the solid overtop placed for blocking sunlight
+	noise_cave1 = new Noise(np_cave1, seed, m_csize.X, m_csize.Y + 1, m_csize.Z);
+	noise_cave2 = new Noise(np_cave2, seed, m_csize.X, m_csize.Y + 1, m_csize.Z);
+}
+
+
+CavesNoiseIntersection::~CavesNoiseIntersection()
 {
-	this->mg             = mg;
-	this->vm             = mg->vm;
-	this->ndef           = mg->ndef;
-	this->water_level    = mg->water_level;
-	this->ps             = ps;
-	c_water_source       = ndef->getId("mapgen_water_source");
-	c_lava_source        = ndef->getId("mapgen_lava_source");
-	c_ice                = ndef->getId("mapgen_ice");
-	this->np_caveliquids = &nparams_caveliquids;
-	this->ystride        = mg->csize.X;
- 
-	if (c_ice == CONTENT_IGNORE)
-		c_ice = CONTENT_AIR;
+	delete noise_cave1;
+	delete noise_cave2;
+}
 
-	dswitchint = ps->range(1, 14);
-	flooded    = ps->range(1, 2) == 2;
 
-	part_max_length_rs  = ps->range(2, 4);
-	tunnel_routepoints  = ps->range(5, ps->range(15, 30));
-	min_tunnel_diameter = 5;
-	max_tunnel_diameter = ps->range(7, ps->range(8, 24));
+void CavesNoiseIntersection::generateCaves(MMVManip *vm,
+	v3s16 nmin, v3s16 nmax, u8 *biomemap)
+{
+	assert(vm);
+	assert(biomemap);
+
+	noise_cave1->perlinMap3D(nmin.X, nmin.Y - 1, nmin.Z);
+	noise_cave2->perlinMap3D(nmin.X, nmin.Y - 1, nmin.Z);
+
+	v3s16 em = vm->m_area.getExtent();
+	u32 index2d = 0;
+
+	for (s16 z = nmin.Z; z <= nmax.Z; z++)
```

Comment:
It's in the rules for nested loops that iterate across coordinates or likewise that they can use this special compressed syntax.  ??

The rationale is that 1). celeron already did that a lot, 2). it makes sense in this case, 3). otherwise there would be too much indentation